Suggestions and Activity Pub spec (#1)
foranetworks/protocol: core/activity_pub.md

# Activity Pub Compliance

This document specifies the extensions to Fora required for a community to be compliant with the [Activity Pub](https://www.w3.org/TR/activitypub/) specification. Without this extension, the Fora kernel will not be capable of interacting with other Activity Pub implementations.

## Introduction

"it's just activity pub with governance"

> If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of users can now collectively agree to be identified as the same community thereby sharing their state and community membership.

Reviewer: ```suggestion
If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```

Lets keep uniform terminology to avoid confusion. When we're talking in ActivityPub context, let's always use the same "Server/Client" pair terminology.

Different servers with different sets of clients may still all be part of the same set of users in a fora community.

The usefulness of this conceptual nesting is addressing the **right to individual exit** and the **right to collective exit**. 

> The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing it's account information (content may be lost, but followers/following lists would not).

Reviewer: ```suggestion
The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).
```

The latter requires the concept of communal ownership, which Activity Pub lacks. Suppose an Activity Pub implementation added support for DID's and thus fulfilled the requirements for the right to individual exit. In the presence of a malicious operator, we can reasonably predict that even in the case of an organized exit, the size and network effect of the forked community will be less than the original community. The right to collective exit and implicitly the right to partial ownership of a community helps mitigates the risk of malicious operators while aligning the incentives of operators, producers, moderators, and consumers.

## Public/Private Key Authentication

> ActivityPub does not specify a specific authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.

Reviewer: ```suggestion
ActivityPub does not specify a particular authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
```

Reviewer: "specify a specific" sounds awkward

> ## URI and DID

Author: This needs a lot more, but I need to dive deeper into dids first

URIs (Uniform Resource Identifier) are used within Activity Pub to identify different servers. 

Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both users and platforms. These DIDs must be registered on a blockchain that the platform can query identities from. Users will have a DID so they can self-attest to their identity on the blockchain and then use that DID in any Fora community they wish. They can also retain this identity even as they move to different communities. The DID document for each user must contain the current public key they will use sign messages sent from their Fora user account. 

Fora implementations which rely on the public key registered with a DID instead of registering a public key with a governance module, will benefit from dynamic swapping of public keys. 

> Meanwhile communities will also have DIDs that identify them to their users. This DID will also be registered on the blockchain as a government. The DID will resolve to the current hosts of the platform. Any governance decisions which affects the set of hosts will update the DID resolving result to output the new set of hosts.

Reviewer: It should be clear that a community does not have a defined set of hosts. Anyone can be a host.

What we want to say here is that a community may define a default set of hosts as a UX convenience.

```suggestion
Meanwhile communities will also have DIDs that identify them to their users. This DID will also be registered on the blockchain as a government. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community. Any governance decisions which affects the set of hosts will update the DID resolving result to output a new set of hosts.

Note: There may be many valid, honest hosts that exist outside of this community-curated set.
```

## Server

> In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.

Reviewer: ```suggestion
In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.
```

## Limitations

> Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though this is grounds for enforcers to remove a host.

Reviewer: ```suggestion
Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.
```

There isn't any absolute way to "remove" a host, bad hosts must simply be circumvented.

# Activity Pub Compliance

This document specifies the extensions to Fora required for a community to be compliant with the [Activity Pub](https://www.w3.org/TR/activitypub/) specification. Without this extension, the Fora kernel will not be capable of interacting with other Activity Pub implementations.

## Introduction

"it's just activity pub with governance"

> If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.

Reviewer: ```suggestion
If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each member is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```

The usefulness of this conceptual nesting is addressing the **right to individual exit** and the **right to collective exit**. 

> The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).

Reviewer: ```suggestion
The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any member the power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).
```

## URI and DID

URIs (Uniform Resource Identifier) are used within Activity Pub to identify different servers and clients.  

Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both members and communities. 

Please refer to the DID specification for further details, but it is essential to understand that:
- a DID is a URI (it can be a URL and it has properties such as Query, Fragment etc)
- a DID identifies a DID Subject, and that subject must never change
- a DID resolves to a DID Document which describes the DID subject
- a DID Document may be modified by one or more DID Controllers
- a DID Subject is not necessarily also the DID Controller
- there exists a verification method for each DID controller, which is specified in the DID Document
- there exists an authentication method to be associated with a DID Subject, which is specified in the DID Document
- All information may be modified in a DID Document except the id of the subject

> The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client (public key the client signs ActivityPub messages with). Communities may choose not to rely on DIDs for authentication of governance module interaction. If the communities choose to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication. It is recommended that governance modules support using a third verification relationship known as `voting authentication` to authenticate any votes casted by a DID controller on behalf of a DID subject. If the `voting authentication` verification relationship does not exist, governance modules should default to using the authroization method of DID controllers to verify casted votes. If communities do not want to require on-chain verifiable DIDs for each member, members must register with the governance module a public key that they will use to cast votes.

Reviewer: ```suggestion
The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client (public key the client signs ActivityPub messages with). Communities may choose not to rely on DIDs for authentication of governance module interaction. If the communities choose to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication. It is recommended that governance modules support using a third verification relationship known as `voting authentication` to authenticate any votes cast by a DID controller on behalf of a DID subject. If the `voting authentication` verification relationship does not exist, governance modules should default to using the authorization method of DID controllers to verify casted votes. If communities do not want to require on-chain verifiable DIDs for each member, members must register with the governance module a public key that they will use to cast votes.
```

Client DIDs are intentionally isolated from the communities they are members of in order to allow for individual exit to different communities as well as improved operational security. DIDs allow for clients to rotate private/public key pairs without notifiying all the relevant parties.  The authentication method specified by a client DID must be a form of public/private key authentication.

> Communities will also have DIDs that identify them to their users. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information. 

Reviewer: ```suggestion
Communities will also have DIDs that identify them to their members. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow members to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information. 
```

Note: There may be many valid, honest hosts that exist outside of this community-curated set.

## Public/Private Key Authentication

ActivityPub does not specify a particular authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives servers privileged information on a client’s authentication info. Thus, clients must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.

## Server

> In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.

Reviewer: ```suggestion
In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of members in the community.
```

## Limitations

> Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.

Reviewer: ```suggestion
Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of members. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to members can be end-to-end encrypted given that all members have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of member-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.
```

The usefulness of this conceptual nesting is addressing the **right to individual exit** and the **right to collective exit**. 

The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).

The latter requires the concept of communal ownership, which Activity Pub lacks. Suppose an Activity Pub implementation added support for DID's and thus fulfilled the requirements for the right to individual exit. In the presence of a malicious operator, we can reasonably predict that even in the case of an organized exit, the size and network effect of the forked community will be less than the original community. The right to collective exit and implicitly the right to partial ownership of a community helps mitigates the risk of malicious operators while aligning the incentives of operators, producers, moderators, and consumers.

## URI and DID

URIs (Uniform Resource Identifier) are used within Activity Pub to identify different servers and clients.  

> Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both clients and communities. 

Reviewer: ```suggestion
Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both members and communities. 
```

Again let's stick to Server/Client for ActivityPub and Community/Member. See #2

Please refer to the DID specification for further details, but it is essential to understand that:
- a DID identifies a DID Subject, and that subject must never change
- a DID resolves to a DID Document which describes the DID subject
- a DID Document may be modified by one or more DID Controllers
- a DID Subject is not necessarily also the DID Controller
- there exists a verification method for each DID controller, which is specified in the DID Document
- there exists an authentication method to be associated with a DID Subject, which is specified in the DID Document
- All information may be modified in a DID Document except the id of the subject

> The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client. Communities may choose not to rely on DIDs for authentication of governance module interaction (NOTE: the authentication method of a client may be different from its corresponding controller, the latter would be used to authenticate a governance vote). In this case, public keys for each client controller must be registered on the community. If the communities chooses to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication.

Reviewer: "(NOTE: the authentication method of a client may be different from its corresponding controller, the latter would be used to authenticate a governance vote)"

Could you clarify what you mean here? Do you mean that the PubKey used to send ActivityPub messages could be different than the pubkey used to send governance votes

Author: Correct. There exists [authorization](https://www.w3.org/TR/did-core/#did-controller) in a DID Document
```
Authorization is the mechanism used to state how operations are performed on behalf of the DID subject. A DID controller is authorized to make changes to the respective DID document. 
```

As well as [authentication](https://www.w3.org/TR/did-core/#authentication). 
```
The authentication verification relationship is used to specify how the DID subject is expected to be authenticated, such as for the purposes of logging into a website. 
```

On further thought, we could just use `authentication` as the method for allowing voting permissions. The controllers could always change the authentication method. 

My concern here is for complex cases, say if a member is actually another fora network. In this case, there may exist a slew of authentication methods (public keys for each sub-member), but authorization might be a multi-sig of the sub-members or even a passed gov proposal. Here authorization is a stronger property than authentication and if the network is to vote, you likely want the strongest verification method?

Thoughts?

Author: Let's do this. Ideally, members have a DID that defines a third verification relationship `voting authentication` or maybe `blockchain authentication`. In this situation, we use that verification relationship for verifying votes. If this verification relationship does not exist (member DIDs could use entirely different DID implementations), then we should gracefully fail on using the authorization method of the DID since I'm pretty sure that should always be there

Author: see pushed changes

Author: I'll put up a new pr later, but looks like we can use capabilities. It's unclear to me rn if you can create multiple capabilities with various delegates for each capability
```
The capabilityDelegation verification relationship is used to specify a mechanism that might be used by the DID subject to delegate a cryptographic capability to another party, such as delegating the authority to access a specific HTTP API to a subordinate.
An example of when this property is useful is when a DID Subject chooses to grant their capability to start a vehicle through the combined usage of a verification method and the StartCar capability to a party other than themselves. 
```

Client DIDs are intentionally isolated from the communities they are members of in order to allow for individual exit to different communities as well as improved operational security. DIDs allow for clients to rotate private/public key pairs without notifiying all the relevant parties.  The authentication method specified by a client DID must be a form of public/private key authentication.

> Communities will also have DIDs that identify them to their users. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information. 

Reviewer: Does the service endpoint only return the list of hosts? Can I use the DID to query for the admittors, enforcers, community parameters, constitution, etc?

Author: Yes, this depends on your DID implementation. [Service endpoints](https://www.w3.org/TR/did-core/#service-endpoints) are natively supported. 

Yes you can, the DID document for the community would include extra [parameters](https://www.w3.org/TR/did-core/#did-parameters) for these keys. Then if you wanted one of those values, you would enter the DID, as a DID URL
`did:cosmoshub5:vov.camp?enforcers=true`

Author: anything I need to add to the doc?
